[REEF-1527] Refactor RuntimeClock and implement graceful shutdown correctly

motus opened the pull request:
Summary of changes:

   * Refactor RuntimeClock for readability and add more Javadoc comments;
   * Fix the bug in `RuntimeClock.close()` that clears all scheduled events without processing them;
   * Clean up and refactor code for Time and derived classes, add Javadocs;
   * Improve logging in RuntimeClock and around;
   * Cosmetic refactoring of RuntimeClock unit tests (new unit tests will be in a separate pull request)

JIRA: [REEF-1527](https://issues.apache.org/jira/browse/REEF-1527) close

markusweimer commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/time/Time.java:
```diff
@@ -18,53 +18,61 @@
 /**
- * Time object.
+ * An abstract object that has a timestamp.
+ * That allows up to compare and order objects by the time.
```
`up` -> `us`

markusweimer commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/time/Time.java:
```diff
@@ -18,53 +18,61 @@
+   * Get timestamp in milliseconds since the beginning of the epoch (01/01/1970).
+   * @return Object's timestamp in milliseconds since the start of the epoch.
+   */
+  public final long getTimestamp() {
```
This is a breaking change. Please keep the old method around and mark it deprecated. File a JIRA to remove it after the next release and mark the method for removal with a `// TODO[REEF-XXX] Remove after release 0.16`

motus replied:
Good point. I've marked the original method as deprecated and opened [REEF-1532](https://issues.apache.org/jira/browse/REEF-1532) in JIRA

markusweimer commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/time/Time.java:
```diff
@@ -18,53 +18,61 @@
   @Override
-  public final String toString() {
-    return this.getClass().getName() + "[" + this.timestamp + "]";
+  public String toString() {
```
Do we have reason to believe that this is a performance bottleneck? If not, why not remove the `timestampStr`?

motus replied:
Conversion `long -> Date -> String` can be pretty expensive, but we don't seem to print events that often. I'll remove the caching logic.

markusweimer commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/time/runtime/LogicalTimer.java:
```diff
@@ -18,36 +18,74 @@
   @Override
-  public long getDuration(final long time) {
-    isReady(time);
+  public long getDuration(final Time time) {
```
Same as above: This is a breaking change, we have to go through the deprecation cycle.

markusweimer commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/time/runtime/LogicalTimer.java:
```diff
@@ -18,36 +18,74 @@
-    if (this.current < time) {
-      this.current = time;
+  public boolean isReady(final Time time) {
+    for (;;) {
```
Is there a way to write this without using `for` in this way? It seems pretty non-idiomatic to me.

motus replied:
It actually *is* the idiomatic way of updating the atomic variable

markusweimer commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/time/runtime/RuntimeClock.java:
```diff
@@ -97,80 +118,107 @@ public void scheduleAlarm(final int offset, final EventHandler<Alarm> handler) {
     synchronized (this.schedule) {
+
+      if (this.isClosed) {
+        LOG.log(Level.INFO, "Clock is already closed");
```
This could become quite spammy on the logs. Maybe move to a more fine-grained log level?

markusweimer commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/time/runtime/RuntimeClock.java:
```diff
@@ -97,80 +118,107 @@ public void scheduleAlarm(final int offset, final EventHandler<Alarm> handler) {
+      this.exceptionCausedStop = exception;
+
+      final Time stopEvent = new StopTime(this.timer.getCurrent());
+      LOG.log(Level.INFO, "Stop scheduled immediately: {0}", stopEvent);
```
Same here, log to `DEBUG` or such.

markusweimer commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/time/runtime/RuntimeClock.java:
```diff
@@ -213,66 +272,73 @@ public void run() {
-      this.handlers.onNext(start);
+      this.handlers.onNext(new StartTime(this.timer.getCurrent()));
+
+      for (;;) {
```
`while (true)` seems more idiomatic to me. What do we gain by this `for` construction?

markusweimer commented:
I've done a pass. LGTM, overall. However, there are a couple of breaking changes in this PR that need to go through the deprecation process.

Markus

motus commented:
Thanks for the thorough code inspection @markusweimer ! I think I've addressed all the issues raised in the review - could you please take another look?

shulmanb commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/time/Time.java:
```diff
@@ -18,53 +18,63 @@
-    return 0;
+  public int compareTo(final Time other) {
+    final int cmp = Long.compare(this.timestamp, other.timestamp);
+    return cmp != 0 ? cmp : Integer.compare(this.hashCode(), other.hashCode());
```
From what I see this.hashCode() will use the default hashCode() implementation, right? So two Time objects will be equal only if they are the same instance. Is this the desired behavior? if yes then wouldn't it be better to implement equals just by comparing references?

markusweimer commented:
LGTM. Will test and merge

shulmanb commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/time/runtime/LogicalTimer.java:
```diff
@@ -18,36 +18,101 @@
+  @Override
+  public long getDuration(final Time time) {
+    return this.getDuration(time.getTimestamp());
+  }
```
Should we rename this method if it always returns 0? Also it will be better to validate that the returned value is actually zero or ignore the returned value and return 0.

shulmanb commented on lang/java/reef-wake/wake/src/main/java/org/apache/reef/wake/time/runtime/LogicalTimer.java:
```diff
@@ -18,36 +18,101 @@
+   */
+  @Override
+  public boolean isReady(final Time time) {
+    return this.isReady(time.getTimestamp());
```
It will be better to validate that the returned value is true, or ignore the returned value and return 'true'

shulmanb commented:
@motus I have done a pass. Overall LGTM.
